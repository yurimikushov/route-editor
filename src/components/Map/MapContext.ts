import { createContext } from 'react'
import { Map } from 'yandex-maps'

const MapContext = createContext<Map | null>(null)

export default MapContext
