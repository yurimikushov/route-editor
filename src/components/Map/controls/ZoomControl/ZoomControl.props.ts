type ZoomControlProps = {
  className?: string
}

export default ZoomControlProps
