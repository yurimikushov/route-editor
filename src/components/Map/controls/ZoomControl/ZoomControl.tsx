import { FC, useContext } from 'react'
import cn from 'classnames'
import isNull from 'lodash/isNull'
import MapContext from '../../MapContext'
import Control from './Control'
import ZoomControlProps from './ZoomControl.props'

const ZoomControl: FC<ZoomControlProps> = ({ className }) => {
  const map = useContext(MapContext)

  if (isNull(map)) {
    throw new Error('[Map] Map should be created before it usage')
  }

  const handleIncreaseZoom = () => {
    map.setZoom(map.getZoom() + 1, {
      duration: 300,
    })
  }

  const handleDecreaseZoom = () => {
    map.setZoom(map.getZoom() - 1, {
      duration: 300,
    })
  }

  return (
    <div
      className={cn(
        className,
        'inline-flex flex-col',
        'bg-white rounded-md',
        'shadow-lg hover:shadow-xl'
      )}
    >
      <Control onClick={handleIncreaseZoom}>+</Control>
      <Control onClick={handleDecreaseZoom}>-</Control>
    </div>
  )
}

export default ZoomControl
