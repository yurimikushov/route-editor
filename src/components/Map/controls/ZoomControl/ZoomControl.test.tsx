import { fireEvent, render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { Map as YMap } from 'yandex-maps'
import MapContext from '../../MapContext'
import ZoomControl from './ZoomControl'

it('should render both + and - controls', () => {
  const tree = renderer
    .create(
      <MapContext.Provider value={{} as YMap}>
        <ZoomControl />
      </MapContext.Provider>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it(`should throw error when map isn't provided`, () => {
  expect(() => render(<ZoomControl />)).toThrowError()
})

it('should increase zoom when + is clicked', () => {
  let currentZoom = 9

  const mockMap = {
    getZoom: jest.fn(() => currentZoom),
    setZoom: jest.fn((zoom: number) => {
      currentZoom = zoom
      return Promise.resolve()
    }),
  } as unknown as YMap

  render(
    <MapContext.Provider value={mockMap}>
      <ZoomControl />
    </MapContext.Provider>
  )

  fireEvent.click(screen.getByText('+'))

  expect(mockMap.getZoom).toHaveBeenCalledTimes(1)
  expect(mockMap.setZoom).toHaveBeenCalledTimes(1)
  expect(mockMap.setZoom).toHaveBeenCalledWith(10, { duration: 300 })

  fireEvent.click(screen.getByText('+'))
  fireEvent.click(screen.getByText('+'))

  expect(mockMap.getZoom).toHaveBeenCalledTimes(3)
  expect(mockMap.setZoom).toHaveBeenCalledTimes(3)
  expect(mockMap.setZoom).toHaveBeenCalledWith(12, { duration: 300 })
  expect(mockMap.setZoom).toHaveBeenCalledWith(12, { duration: 300 })
})

it('should decrease zoom when - is clicked', () => {
  let currentZoom = 9

  const mockMap = {
    getZoom: jest.fn(() => currentZoom),
    setZoom: jest.fn((zoom: number) => {
      currentZoom = zoom
      return Promise.resolve()
    }),
  } as unknown as YMap

  render(
    <MapContext.Provider value={mockMap}>
      <ZoomControl />
    </MapContext.Provider>
  )

  fireEvent.click(screen.getByText('-'))

  expect(mockMap.getZoom).toHaveBeenCalledTimes(1)
  expect(mockMap.setZoom).toHaveBeenCalledTimes(1)
  expect(mockMap.setZoom).toHaveBeenCalledWith(8, { duration: 300 })

  fireEvent.click(screen.getByText('-'))
  fireEvent.click(screen.getByText('-'))

  expect(mockMap.getZoom).toHaveBeenCalledTimes(3)
  expect(mockMap.setZoom).toHaveBeenCalledTimes(3)
  expect(mockMap.setZoom).toHaveBeenCalledWith(7, { duration: 300 })
  expect(mockMap.setZoom).toHaveBeenCalledWith(6, { duration: 300 })
})
