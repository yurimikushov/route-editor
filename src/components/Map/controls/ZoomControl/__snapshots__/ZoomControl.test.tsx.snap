// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render both + and - controls 1`] = `
<div
  className="inline-flex flex-col bg-white rounded-md shadow-lg hover:shadow-xl"
>
  <button
    className="px-3 py-2.5 rounded-md text-2xl text-gray-500 leading-none hover:text-black focus:outline-none focus-visible:ring-2 focus-visible:ring-indigo-500"
    onClick={[Function]}
  >
    +
  </button>
  <button
    className="px-3 py-2.5 rounded-md text-2xl text-gray-500 leading-none hover:text-black focus:outline-none focus-visible:ring-2 focus-visible:ring-indigo-500"
    onClick={[Function]}
  >
    -
  </button>
</div>
`;
